### PR TITLE
minor refactoring of ingress v2 lib

### PR DESCRIPTION
**Pull request by PietroPasotti**
added some tests to verify the behaviour of ingress v2, namely:

- a smooth migration from v1 to v2 on upgrade
- cleanup on relation remove of config files from the filesystem

minor refactoring of ingress v2 lib code for readability and maintainability

- fixed a bug in `.load()`, where a raw jsondecodeerror was raised instead of wrapping it in DataValidationError, as expected by the caller

### CODE REVIEW

**Review by sed-i — COMMENTED** [review]
Hey @PietroPasotti,
It's not easy to predict consequences these days :)
If not already in itests, any chance you could manually test:
- when deployed in a bundle like cos-lite, remove only traefik
- when deployed in a bundle like cos-lite, remove all apps except traefik

**Comment by gboutry** [comment]
Hi, is there a way to get this moving forward ?

We would like to migrate all Sunbeam charms to V2 ingress to get load balancing across units, and this PR fixes our issue with updating charms in-place.

File: tests/scenario/test_ingress_per_app.py [hunk 1]
```diff
@@ -167,3 +176,69 @@ def __init__(self, framework: Framework):
             "scheme": '"http"',
             "strip-prefix": "false",
         }
+
+
+def test_ingress_per_app_cleanup_on_remove(model, traefik_ctx, traefik_container):
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply tests/scenario/test_ingress_per_app.py:181]: I'm not sure I understand the concern. 
In what way is the initial condition not matching 'reality'?
Or is the setup just too complex to fully grasp?

**Review by lucabello — APPROVED** [review]

File: tests/scenario/test_ingress_per_app.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply tests/scenario/test_ingress_per_app.py:181]: tested both 'remove all but traefik' and 'remove traefik' paths you suggested, traefik ends up in active/idle, no errors I could see

[the remark this answers is not in the document]
Comment (sed-i) [reply tests/scenario/test_ingress_per_app.py:181]: Yeah the prep step is a bit hard to grasp.

File: lib/charms/traefik_k8s/v2/ingress.py [hunk 2]
```diff
@@ -120,8 +121,15 @@ def load(cls, databag: MutableMapping):
...
-    def dump(self, databag: Optional[MutableMapping] = None):
-        """Write the contents of this model to Juju databag."""
+    def dump(self, databag: Optional[MutableMapping] = None, clear: bool = True):
+        """Write the contents of this model to Juju databag.
+
+        :param databag: the databag to write the data to.
+        :param clear: ensure the databag is cleared before writing it.
+        """
+        if clear and databag:
+            databag.clear()
```
Comment (sed-i) [inline lib/charms/traefik_k8s/v2/ingress.py:131]: (In a future PR) we should rethink the `clear`. It seems un-jujuy to do so. 
- ops itself prevents assignment of relation data
- relation mgmt libs is our own invention; nothing prevents charm code or other libs to write to the same relation data

**Review by sed-i — APPROVED** [review]

File: lib/charms/traefik_k8s/v2/ingress.py (hunk 2 above)
Comment (PietroPasotti) [reply lib/charms/traefik_k8s/v2/ingress.py:131]: `.defer()` --> https://github.com/canonical/traefik-k8s-operator/issues/252

**Comment by PietroPasotti** [comment]
@gboutry stay tuned for a release on edge as soon as the CI is done

**Comment by gboutry** [comment]
Thank you very much!

**Pull request merged**